### PR TITLE
optimize kubelet client certificate manager: use async notification

### LINKED ISSUE #52622: kubelet ClientCertificateManager should use async notification rather than periodically checks to know certificate rotation
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
> /kind feature

**What happened**:
Currently kubelet client certificate manager use [periodically checks ](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/certificate/transport.go#L53-L113) to check kubelet client certificate rotaion.
**What you expected to happen**:
I think we can optimize this mechanism, by asynchronous channel notification when real rotateCerts happens.
**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration**:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #52622 

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/kubelet/certificate/transport.go [hunk 1]
```diff
@@ -91,7 +93,7 @@ func updateTransport(stopCh <-chan struct{}, period time.Duration, clientConfig
 		//
 		// See: https://github.com/kubernetes-incubator/bootkube/pull/663#issuecomment-318506493
 		t.closeAllConns()
-	}, period, stopCh)
+	}, 0, stopCh)
```
Comment (liggitt) [inline pkg/kubelet/certificate/transport.go]: no 0 waits... that's a hotloop if the rotate channel gets closed

File: pkg/kubelet/certificate/transport_test.go [hunk 2]
```diff
@@ -106,6 +107,17 @@ func (f *fakeManager) Current() *tls.Certificate {
 	return nil
 }
 
+func (f *fakeManager) RotateCh() <-chan bool {
+	return f.rotateCh
+}
+
+func (f *fakeManager) T() *tls.Certificate {
```
Comment (liggitt) [inline pkg/kubelet/certificate/transport_test.go]: what is this for?

File: pkg/kubelet/certificate/certificate_manager.go [hunk 3]
```diff
@@ -55,6 +55,8 @@ type Manager interface {
 	// certificate manager, as well as the associated certificate and key data
 	// in PEM format.
 	Current() *tls.Certificate
+	// RotateCh returns the certificate rotate channel.
+	RotateCh() <-chan bool
```
Comment (liggitt) [inline pkg/kubelet/certificate/certificate_manager.go]: a channel here only allows for a single consumer, which doesn't seem right. do you want a sync.Cond instead?

File: pkg/kubelet/certificate/transport_test.go (hunk 2 above)
Comment (hzxuzhonghu) [reply pkg/kubelet/certificate/transport_test.go]: Sorry， I need remove T()

File: pkg/kubelet/certificate/transport.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pkg/kubelet/certificate/transport.go]: the channel should not be closed in any case by hand. Do you mean need process exceptional close here?

File: pkg/kubelet/certificate/certificate_manager.go (hunk 3 above)
Comment (hzxuzhonghu) [reply pkg/kubelet/certificate/certificate_manager.go]: One consumer channel not wright? I donot think so. Do you think sync.Cond is better than channel here?

Comment (hzxuzhonghu) [reply pkg/kubelet/certificate/certificate_manager.go]: @liggitt OK. I think about using sync.Cond here. Maybe better

**Comment by hzxuzhonghu** [comment]
@liggitt Done, use sync.Cond

**Comment by smarterclayton** [comment]
Is this really worth the complexity?  What sort of scenario does this become required?

**Comment by hzxuzhonghu** [comment]
@smarterclayton  poll check is a waste of cpu resource and also has average 1/2 poll period deplay.

**Comment by hzxuzhonghu** [comment]
kindly ping @liggitt @jcbsmpsn @mikedanese would you please take a look?

**Comment by smarterclayton** [comment]
We know exactly when the certificate is going to expire, why is the poll period inefficient?

**Comment by hzxuzhonghu** [comment]
@smarterclayton Yes, periodically check is ok. but as we can know when the certificate is rotated, why do not use async notification, which can reduce resource waste. Although periodically check can work well in this case, but async notification is more graceful and widely used to implement high performance.

**Comment by smarterclayton** [comment]
If we know the period up front, the timed code is simpler to support
and maintain.

Re: performance I don't even think you could measure a difference with
this change (but if you have data supporting that I'd be happy to
look).  This is a tiny fraction of the Kubelet workload and so this
just seems like a premature optimization.  If you can point to a
performance impact of this I'm happy to look again.

**Comment by hzxuzhonghu** [comment]
@smarterclayton I agree the effect of performance is tiny. 
>If we know the period up front, the timed code is simpler to support
and maintain.
As to this,  kubelet transport does not record the certificate expire deadline, so it periodically check and compare. The period is set arbitrary., why not try a aync mechanism. I still can not understand what you said complexity. By the way, if all resources use periodically poll (kubelet watch pod, why not poll), the cluster performance is terrible. So I think we should also pursue outstanding mechanism.

**Comment by mikedanese** [comment]
This seems like an unnecessary optimization

**Comment by hzxuzhonghu** [comment]
@mikedanese I donot think so. If you think async notification is unnecessary, why not kubeClient periodically poll apiserver?

**Comment by liggitt** [comment]
> @mikedanese I donot think so. If you think async notification is unnecessary, why not kubeClient periodically poll apiserver?

There's a continuous stream of updates coming from the API server, and polling is very costly.

Polling here is very cheap (probably cheaper than parsing the cert a single time)

I agree that this optimization is not worth the added complexity.

**Pull request closed without merging**